helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedInstanceCapacity.java

```diff
@@ -54,7 +55,15 @@ public WagedInstanceCapacity(ResourceControllerDataProvider clusterData) {
       return;
     }
     for (InstanceConfig instanceConfig : clusterData.getInstanceConfigMap().values()) {
-      Map<String, Integer> instanceCapacity = WagedValidationUtil.validateAndGetInstanceCapacity(clusterConfig, instanceConfig);
+      Map<String, Integer> instanceCapacity;
+      try {
+        instanceCapacity = WagedValidationUtil.validateAndGetInstanceCapacity(clusterConfig, instanceConfig);
+      } catch (HelixException ex) {
+        // We don't want to throw exception here, it would be OK if no resource is using Waged.
+        // Waged rebalancer will fail in later pipeline stage only for waged resource. So it won't block other resources.
+        LOG.error("Failed to initialize instance capacity map. Instance capacity map is not property set up", ex);
+        return;
```

Comment:
I don't think that it is safe to return here. We can catch the exception but we need to add empty InstanceCapacity map to  _instanceCapacityMap for each instance.

Consider the following case:
- Cluster does not use DEFAULT_INSTANCE_CAPACITY_MAP but they populate the INSTANCE_CONFIG_WEIGHTS at the InstanceConfig level.
- Let's assume they also have WAGED resources
- Everything is working as expected
- Now lets assume that somehow the INSTANCE_CAPACITY_MAP gets deleted from one InstanceConfig
- If we exit this method early without adding all of the instances to _instanceCapacityMap, we will get null pointer exception [here](https://github.com/apache/helix/blob/master/helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedInstanceCapacity.java#L192) because we are trying to charge capacity [here](https://github.com/apache/helix/blob/master/helix-core/src/main/java/org/apache/helix/controller/rebalancer/DelayedAutoRebalancer.java#L382) for instances that may have not been added before returning.
- I don't think it will be an issue for the instance that is missing INSTANCE_CAPACITY_MAP because all assignments to that instance will be moved since there is no capacity; therefore, it will never be charged.

Reply:
After looking at AssignableNode creation logic, it looks like there will be an exception thrown when creating the ClusterModel for WAGED as `validateAndGetInstanceCapacity` is called for that. This would mean that we have a pipeline failure before we start to charge _instanceCapacityMap.

That being said, I am not sure if we want to fail the pipeline in this situation. Would it not be better to default required capacity keys to 0 if nothing is present in InstanceConfig INSTANCE_CAPACITY_MAP or ClusterConfig DEFAULT_INSTANCE_CAPACITY_MAP?

If we did that, we would not impact non-waged resources by failing the whole pipeline and would also be able to continue assigning to other nodes with the correct configurations.

Reply:
+1 here. This is under the assumption that if the instanceCapacity is not available then `getInstanceAvailableCapacity`, `isInstanceCapacityAvailable`, and `checkAndReduceInstanceCapacity` won't be called. It's risky. Could replace all `_instanceCapacityMap.get()` with `getOrDefault()` to give instances that doesn't exist in INSTANCE_CAPACITY_MAP a default value of 0. In this way, if the INSTANCE_CAPACITY_MAP is not available, no WAGED-related resources can allocate partition on this instance.

Reply:
Thanks for your review.
We would like to keep the previous behavior. Because Waged is doing global capacity planning, we dont want to assign any waged related resource when there is at least one instance without configured capacity.

Reply:
For our current known waged users, they do not have mixed crushed and waged usage. So it wont cause any NPR error. However, I do agree we should put empty map to avoid NPR. I will update.